homeassistant/components/sensor/phicomm.py

```diff
@@ -0,0 +1,196 @@
+}
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_DEVICES, default=1): cv.positive_int,
+    vol.Optional(CONF_SENSORS, default=DEFAULT_SENSORS):
+        vol.All(cv.ensure_list, vol.Length(min=1), [vol.In(SENSOR_MAP)]),
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Phicomm sensor."""
+    name = config.get(CONF_NAME)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    count = config.get(CONF_DEVICES)
+    sensors = config[CONF_SENSORS]
+
+    phicomm = PhicommData(hass.config.path(TOKEN_FILE + username),
+                          username, password, count * len(sensors))
+
+    index = 0
+    devices = []
+    while index < count:
+        for sensor_type in sensors:
+            devices.append(PhicommSensor(phicomm, name, index, sensor_type))
+        index += 1
+    add_devices(devices, True)
+
+
+class PhicommSensor(Entity):
+    """Implementation of a Phicomm sensor."""
+
+    def __init__(self, phicomm, name, index, sensor_type):
+        """Initialize the Phicomm sensor."""
+        sensor_name, unit, icon = SENSOR_MAP[sensor_type]
+        if index:
+            name += str(index + 1)
+        self._name = name + ' ' + sensor_name
+        self._index = index
+        self._sensor_type = sensor_type
+        self._unit = unit
+        self._icon = 'mdi:' + icon
+        self.phicomm = phicomm
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon of the sensor."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit
+
+    @property
+    def available(self):
+        """Return if the sensor data are available."""
+        data = self.data
+        return data and data.get('online') == '1'
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        data = self.data
+        return data.get(self._sensor_type) if data else None
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self.data if self._sensor_type == SENSOR_PM25 else None
+
+    def update(self):
+        """Get the latest data from Phicomm server and update the state."""
+        _LOGGER.info("Begin update: %s", self.name)
+        self.phicomm.update()
+
+    @property
+    def data(self):
+        """Get data with current device index."""
+        devs = self.phicomm.devs
+        if devs and self._index < len(devs):
+            return devs[self._index].get('catDev')
+        return None
+
+
+class PhicommData():
```

Comment:
This should be moved to an external package and hosted on pypi.
With little effort this could also be made async,